gum/backend-qnx/gummemory-qnx.c

```diff
@@ -0,0 +1,216 @@
+        (prot != NULL) ? prot : &ignored_prot);
+  }
+
+  if (n > 1)
+  {
+    GumAddress page_size, start_page, end_page, cur_page;
+
+    page_size = gum_query_page_size ();
+
+    start_page = address & ~(page_size - 1);
+    end_page = (address + n - 1) & ~(page_size - 1);
+
+    success = gum_memory_get_protection (start_page, 1, NULL, prot);
+    if (success)
+    {
+      *size = page_size - (address - start_page);
+      for (cur_page = start_page + page_size;
+          cur_page != end_page + page_size;
+          cur_page += page_size)
+      {
+        GumPageProtection cur_prot;
+
+        if (gum_memory_get_protection (cur_page, 1, NULL, &cur_prot)
+            && (cur_prot != GUM_PAGE_NO_ACCESS || *prot == GUM_PAGE_NO_ACCESS))
+        {
+          *size += page_size;
+          *prot &= cur_prot;
+        }
+        else
+        {
+          break;
+        }
+      }
+      *size = MIN (*size, n);
+    }
+
+    return success;
+  }
+
+
+  success = FALSE;
+  *size = 0;
+  *prot = GUM_PAGE_NO_ACCESS;
+
+  fd = open ("/proc/self/as", O_RDONLY);
+  g_assert (fd != -1);
+
+  res = devctl (fd, DCMD_PROC_PAGEDATA, 0, 0, &num_mapinfos);
+  g_assert (res == 0);
+
+  mapinfos = g_malloc (sizeof (procfs_mapinfo) * num_mapinfos);
+
+  res = devctl (fd, DCMD_PROC_PAGEDATA, mapinfos,
+      sizeof (procfs_mapinfo) * num_mapinfos, &num_mapinfos);
+  g_assert (res == 0);
+
+  while (i != num_mapinfos)
+  {
+    start = GSIZE_TO_POINTER (mapinfos[i].vaddr & 0xffffffff);
+    end = start + mapinfos[i].size;
+
+    if (GUM_ADDRESS (start) > address)
+      break;
+    else if (address >= GUM_ADDRESS (start) &&
+        address + n -1 < GUM_ADDRESS (end))
+    {
+      success = TRUE;
+      *size = 1;
+
+      *prot = _gum_page_protection_from_posix (mapinfos[i].flags);
+      break;
+    }
+
+    i++;
+  }
+
+  close (fd);
+  g_free (mapinfos);
+  return success;
+}
+
+
+void
+gum_clear_cache (gpointer address,
+                 gsize size)
+{
+  g_assert_not_reached ();
+}
+
+gboolean
+gum_memory_is_readable (GumAddress address,
+                        gsize len)
+{
+  gsize size;
+  GumPageProtection prot;
+
+  if (!gum_memory_get_protection (address, len, &size, &prot))
+    return FALSE;
+
+  return size >= len && (prot & GUM_PAGE_READ) != 0;
+}
+
+static gboolean
+gum_memory_is_writable (GumAddress address,
+                        gsize len)
+{
+  gsize size;
+  GumPageProtection prot;
+
+  if (!gum_memory_get_protection (address, len, &size, &prot))
+    return FALSE;
+
+  return size >= len && (prot & GUM_PAGE_WRITE) != 0;
+}
+
+guint8 *
+gum_memory_read (GumAddress address,
+                 gsize len,
+                 gsize * n_bytes_read)
+{
+  FILE *fp = NULL;
+  guint8 * buffer = NULL;
+  gint num_read = 0;
+  gint res = 0;
+
+  fp = fopen("/proc/self/as", "r");
+  res = fseek (fp, address, SEEK_SET);
+  g_assert (res == 0);
+
+  buffer = g_malloc (len);
+  num_read = fread (buffer, 1, len, fp);
+  if (!num_read) {
```

Comment:
We consistently prefer explicit checks, so instead of:
```C
if (!num_read)
{
}
```

we do:

```C
if (num_read == 0)
{
}
```

The only exception is booleans.